File: client/python/request.py
from enum import Enum, auto
from typing import Optional

from addict import Dict
from bson import encode


class Action(Enum):
    create = auto()
    retrieve = auto()
    update = auto()
    delete = auto()
    count = auto()
    index = auto()
    dropCollection = auto()
    dropIndex = auto()
    bulk = auto()
    pipeline = auto()
    transaction = auto()
    renameCollection = auto()


class Request:
    def __init__(self, database: str, collection: str, document: Dict, action: Action,
                 options: Optional[Dict] = None, metadata: Optional[Dict] = None,
                 correlation_id: str = "", skip_version: bool = False,
                 skip_metric: bool = False):
        self._database = database
        self._collection = collection
        self._document = document
        self._action = action
        self._options = options
        self._metadata = metadata
        self._correlation_id = correlation_id
        self._skip_version = skip_version
        self._skip_metric = skip_metric

    def bson(self) -> bytes:
        d = Dict()
        d.database = self._database
        d.collection = self._collection
        d.document = self._document
        d.action = self._action.name

        if self._options:
            d.options = self._options
        if self._metadata:
            d.metadata = self._metadata
        if self._correlation_id:
            d.correlationId = self._correlation_id
        if self._skip_version:
            d.skipVersion = self._skip_version
        if self._skip_metric:
            d.skipMetric = self._skip_metric

        return encode(d)


def create_request(database: str, collection: str, document: Dict, options: Optional[Dict] = None,
                   metadata: Optional[Dict] = None, correlation_id: str = "", skip_version: bool = False,
                   skip_metric: bool = False) -> Request:
    return Request(database=database, collection=collection, document=document, action=Action.create,
                   options=options, metadata=metadata, correlation_id=correlation_id, skip_version=skip_version,
                   skip_metric=skip_metric)


def retrieve_request(database: str, collection: str, document: Dict, options: Optional[Dict] = None,
                     metadata: Optional[Dict] = None, correlation_id: str = "", skip_version: bool = False,
                     skip_metric: bool = False) -> Request:
    return Request(database=database, collection=collection, document=document, action=Action.retrieve,
                   options=options, metadata=metadata, correlation_id=correlation_id, skip_version=skip_version,
                   skip_metric=skip_metric)


def update_request(database: str, collection: str, document: Dict, options: Optional[Dict] = None,
                   metadata: Optional[Dict] = None, correlation_id: str = "", skip_version: bool = False,
                   skip_metric: bool = False) -> Request:
    return Request(database=database, collection=collection, document=document, action=Action.update,
                   options=options, metadata=metadata, correlation_id=correlation_id, skip_version=skip_version,
                   skip_metric=skip_metric)


def delete_request(database: str, collection: str, document: Dict, options: Optional[Dict] = None,
                   metadata: Optional[Dict] = None, correlation_id: str = "", skip_version: bool = False,
                   skip_metric: bool = False) -> Request:
    return Request(database=database, collection=collection, document=document, action=Action.delete,
                   options=options, metadata=metadata, correlation_id=correlation_id, skip_version=skip_version,
                   skip_metric=skip_metric)


def count_request(database: str, collection: str, options: Optional[Dict] = None,
                  metadata: Optional[Dict] = None, correlation_id: str = "",
                  skip_metric: bool = False) -> Request:
    return Request(database=database, collection=collection, document=Dict(), action=Action.count,
                   options=options, metadata=metadata, correlation_id=correlation_id,
                   skip_metric=skip_metric)


def index_request(database: str, collection: str, document: Dict, options: Optional[Dict] = None,
                  metadata: Optional[Dict] = None, correlation_id: str = "", skip_metric: bool = False) -> Request:
    return Request(database=database, collection=collection, document=document, action=Action.index,
                   options=options, metadata=metadata, correlation_id=correlation_id, skip_metric=skip_metric)


def drop_index_request(database: str, collection: str, document: Dict, options: Optional[Dict] = None,
                       metadata: Optional[Dict] = None, correlation_id: str = "", skip_metric: bool = False) -> Request:
    return Request(database=database, collection=collection, document=document, action=Action.dropIndex,
                   options=options, metadata=metadata, correlation_id=correlation_id, skip_metric=skip_metric)


def drop_collection_request(database: str, collection: str, document: Dict, options: Optional[Dict] = None,
                            metadata: Optional[Dict] = None, correlation_id: str = "",
                            skip_metric: bool = False) -> Request:
    return Request(database=database, collection=collection, document=document, action=Action.dropCollection,
                   options=options, metadata=metadata, correlation_id=correlation_id, skip_metric=skip_metric)


def bulk_request(database: str, collection: str, document: Dict, options: Optional[Dict] = None,
                 metadata: Optional[Dict] = None, correlation_id: str = "", skip_version: bool = False,
                 skip_metric: bool = False) -> Request:
    return Request(database=database, collection=collection, document=document, action=Action.bulk,
                   options=options, metadata=metadata, correlation_id=correlation_id, skip_version=skip_version,
                   skip_metric=skip_metric)


def pipeline_request(database: str, collection: str, document: Dict, options: Optional[Dict] = None,
                     metadata: Optional[Dict] = None, correlation_id: str = "", skip_version: bool = False,
                     skip_metric: bool = False) -> Request:
    return Request(database=database, collection=collection, document=document, action=Action.pipeline,
                   options=options, metadata=metadata, correlation_id=correlation_id, skip_version=skip_version,
                   skip_metric=skip_metric)


def transaction_request(database: str, collection: str, document: Dict, options: Optional[Dict] = None,
                        metadata: Optional[Dict] = None, correlation_id: str = "",
                        skip_version: bool = False, skip_metric: bool = False) -> Request:
    return Request(database=database, collection=collection, document=document, action=Action.transaction,
                   options=options, metadata=metadata, correlation_id=correlation_id, skip_version=skip_version,
                   skip_metric=skip_metric)


def rename_collection_request(database: str, collection: str, document: Dict, options: Optional[Dict] = None,
                              correlation_id: str = "", skip_metric: bool = False) -> Request:
    return Request(database=database, collection=collection, document=document, action=Action.renameCollection,
                   options=options, metadata=None, correlation_id=correlation_id, skip_version=False,
                   skip_metric=skip_metric)

File: client/julia/MongoService/src/request.jl
import LightBSON: bson_write

module Actions
@enum(Action::UInt8,
    create, retrieve, update, delete, count,
    index, dropCollection, dropIndex,
    bulk, pipeline, transaction
)
end

const Document = Dict{String,Any}

struct Request
    database::String
    collection::String
    document::Document
    action::Actions.Action
    options::Document
    metadata::Document
    correlationId::String
    skipVersion::Bool
    skipMetric::Bool
end

Request(;database::String, collection::String, document::Document, action::Actions.Action,
    options::Document = Document(), metadata::Document = Document(),
    correlationId::String = "", skipVersion::Bool = false, skipMetric::Bool = false ) =
    Request(database, collection, document, action, options, metadata, correlationId, skipVersion, skipMetric)

CreateRequest(;database::String, collection::String, document::Document,
    options::Document = Document(), metadata::Document = Document(),
    correlationId::String = "", skipVersion::Bool = false, skipMetric::Bool = false ) =
    Request(database, collection, document, Actions.create, options, metadata, correlationId, skipVersion, skipMetric)

RetrieveRequest(;database::String, collection::String, document::Document,
    options::Document = Document(), metadata::Document = Document(),
    correlationId::String = "", skipVersion::Bool = false, skipMetric::Bool = false ) =
    Request(database, collection, document, Actions.retrieve, options, metadata, correlationId, skipVersion, skipMetric)

UpdateRequest(;database::String, collection::String, document::Document,
    options::Document = Document(), metadata::Document = Document(),
    correlationId::String = "", skipVersion::Bool = false, skipMetric::Bool = false ) =
    Request(database, collection, document, Actions.update, options, metadata, correlationId, skipVersion, skipMetric)

DeleteRequest(;database::String, collection::String, document::Document,
    options::Document = Document(), metadata::Document = Document(),
    correlationId::String = "", skipVersion::Bool = false, skipMetric::Bool = false ) =
    Request(database, collection, document, Actions.delete, options, metadata, correlationId, skipVersion, skipMetric)

CountRequest(;database::String, collection::String,
    options::Document = Document(), metadata::Document = Document(),
    correlationId::String = "", skipVersion::Bool = false, skipMetric::Bool = false ) =
    Request(database, collection, Document(), Actions.count, options, metadata, correlationId, skipVersion, skipMetric)

IndexRequest(;database::String, collection::String, document::Document,
    options::Document = Document(), metadata::Document = Document(),
    correlationId::String = "", skipVersion::Bool = false, skipMetric::Bool = false ) =
    Request(database, collection, document, Actions.index, options, metadata, correlationId, skipVersion, skipMetric)

DropCollectionRequest(;database::String, collection::String, document::Document,
    options::Document = Document(), metadata::Document = Document(),
    correlationId::String = "", skipVersion::Bool = false, skipMetric::Bool = false ) =
    Request(database, collection, document, Actions.dropCollection, options, metadata, correlationId, skipVersion, skipMetric)

DropIndexRequest(;database::String, collection::String, document::Document,
    options::Document = Document(), metadata::Document = Document(),
    correlationId::String = "", skipVersion::Bool = false, skipMetric::Bool = false ) =
    Request(database, collection, document, Actions.dropIndex, options, metadata, correlationId, skipVersion, skipMetric)

BulkRequest(;database::String, collection::String, document::Document,
    options::Document = Document(), metadata::Document = Document(),
    correlationId::String = "", skipVersion::Bool = false, skipMetric::Bool = false ) =
    Request(database, collection, document, Actions.bulk, options, metadata, correlationId, skipVersion, skipMetric)

PipelineRequest(;database::String, collection::String, document::Document,
    options::Document = Document(), metadata::Document = Document(),
    correlationId::String = "", skipVersion::Bool = false, skipMetric::Bool = false ) =
    Request(database, collection, document, Actions.pipeline, options, metadata, correlationId, skipVersion, skipMetric)

TransactionRequest(;database::String, collection::String, document::Document,
    options::Document = Document(), metadata::Document = Document(),
    correlationId::String = "", skipVersion::Bool = false, skipMetric::Bool = false ) =
    Request(database, collection, document, Actions.transaction, options, metadata, correlationId, skipVersion, skipMetric)

function to_bson(r::Request, application::String)::Vector{UInt8}
    buf = Vector{UInt8}()
    d = Dict("database" => r.database, "collection" => r.collection,
        "document" => r.document, "application" => application)
    if r.action == Actions.create
        d["action"] = "create"
    elseif r.action == Actions.retrieve
        d["action"] = "retrieve"
    elseif r.action == Actions.update
        d["action"] = "update"
    elseif r.action == Actions.delete
        d["action"] = "delete"
    elseif r.action == Actions.count
        d["action"] = "count"
    elseif r.action == Actions.index
        d["action"] = "index"
    elseif r.action == Actions.dropCollection
        d["action"] = "dropCollection"
    elseif r.action == Actions.dropIndex
        d["action"] = "dropIndex"
    elseif r.action == Actions.bulk
        d["action"] = "bulk"
    elseif r.action == Actions.pipeline
        d["action"] = "pipeline"
    elseif r.action == Actions.transaction
        d["action"] = "transaction"
    end

    if length(r.options) > 0 d["options"] = r.options end
    if length(r.metadata) > 0 d["metadata"] = r.metadata end
    if length(r.correlationId) > 0 d["correlationId"] = r.correlationId end
    if r.skipVersion d["skipVersion"] = true end
    if r.skipMetric d["skipMetric"] = true end

    bson_write(buf, d)
    return buf
end
